fix JSON schema with MLM fields + support pydantic/pystac objects (#2)
rbavery/dlm-extension: README.md

| `regression`            | `regression`                | Generic regression that estimates a numeric and continuous value.                                               |
| `classification`        | `classification`            | Generic classification task that assigns class labels to an output.                                             |
| `scene-classification`  | *n/a*                       | Specific classification task where the model assigns a single class label to an entire scene/area.              |
| `detection`             | `detection`                 | Generic detection of the "presence" of objects or entities, with or without positions.                          |
| `object-detection`      | *n/a*                       | Task corresponding to the identification of positions as bounding boxes of object detected in the scene.        |
| `segmentation`          | `segmentation`              | Generic tasks that regroups all types of segmentations tasks consisting of applying labels to pixels.           |

> | `semantic-segmentation` | *n/a*                       | Specific segmentation task where all pixels are attributed labels, without consideration of similar instances.  |

Reviewer: ```suggestion
| `semantic-segmentation` | *n/a*                       | Specific segmentation task where all pixels are attributed labels, without consideration for segments as unique objects.  |
```

this should align with the `label:tasks` values defined in [STAC Label Extension][stac-ext-label-props] for relevant
STAC Collections and Items published with the model described by this extension.

It is to be noted that multiple "*generic*" tasks names (`classification`, `detection`, etc.) are defined to allow
correspondance with `label:tasks`, but these can lead to some ambiguity depending on context. For example, a model
that supports `classification` could mean that the model can predict patch-based classes over an entire scene
(i.e.: `scene-classification` for a single prediction over an entire area of interest as a whole),
or that it can predict pixel-wise "classifications", such as land-cover labels for
every single pixel coordinate over the area of interest. Maybe counter-intuitively to some users,
such a model that produces pixel-wise "classifications" should be attributed the `segmentation` task
(and more specifically `semantic-segmentation`) rather than `classification`. To avoid this kind of ambiguity,
it is strongly recommended that `tasks` always aim to provide the most specific definitions possible to explicitly

> describe what the model accomplishes.

Reviewer: great explanation here!

[stac-ext-label-props]: https://github.com/stac-extensions/label#item-properties

### Framework

> In most cases, this should correspond to common library names of well-established ML frameworks.

Reviewer: ```suggestion
This should correspond to the common library name of the well-established ML framework used to train the model.
```

No explicit "Enum" is defined to allow easy addition of newer frameworks, but it is recommended
to use common names when applicable. Below are a few notable entries.

- `PyTorch`
- `TensorFlow`
- `Scikit-learn`
- `Huggingface`
- `PyMC`
- `JAX`
- `ONNX`
- `MXNet`
- `Keras`
- `Caffe`

> - `Weka`

Reviewer: should `Weka` be listed? I've never heard of it or seen it in the wild.

Here's a suggested reordering based on my subjective interpretation of current popularity + longevity. 

I also added `rgee` and `spatialRF` to showcase some R options. especially in academia, lots of folks use R, particularly random forest models for semantic segmentation.

I removed Caffe (no updates in 4 years) and MxNet (archived last year). I don't think anyone will publish models for these frameworks.

Removed ONNX since it isn't a training framework and I think the purpose of this field is to describe the framework used to train the model. this might be different than the inference runtime and format.

```suggestion
- `PyTorch`
- `TensorFlow`
- `Scikit-learn`
- `Huggingface`
- `Keras`
- `rgee`
- `spatialRF`
- `JAX`
- `PyMC`
```

- `inf`
- `clip`

See [OpenCV - Interpolation Flags](https://docs.opencv.org/4.x/da/d54/group__imgproc__transform.html#ga5bb5a1fea74ea38e1a5445ca803ff121)
for details about the relevant methods. Equivalent methods from other packages are applicable as well.

When a normalization technique is specified, it is expected that the corresponding [Statistics](#bands-and-statistics)
parameters necessary to perform it would be provided for the corresponding input.
For example, the `min-max` normalization would require that at least the `minimum` and `maximum` statistic properties
are provided, while the `z-score` would require `mean` and `stddev`.

If none of the above values applies, `null` (literal, not string) can be used instead.
If a custom normalization operation, or a combination of operations (with or without [Resize](#resize-enum)),
must be defined instead, consider using a [Processing Expression](#processing-expression) reference.

#### Resize Enum

Select one option from:
- `crop`
- `pad`
- `interpolation-nearest`
- `interpolation-linear`
- `interpolation-cubic`
- `interpolation-area`
- `interpolation-lanczos4`
- `interpolation-max`
- `wrap-fill-outliers`
- `wrap-inverse-map`

> See [OpenCV - Normalization Flags](https://docs.opencv.org/4.x/d2/de8/group__core__array.html#ga87eef7ee3970f86906d69a92cbf064bd)

Reviewer: I think this reference to normalization flags needs to be switched with the reference to interpolation.

long term I'd be interested in picking a different reference than OpenCV's C++ documentation, since the OpenCV lib is lower level than most folks encounter and the docs  are a bit hard to follow (python programmer might get confused with the C data types for example). But I think this is better than us rolling our own.

| type              | string                                    | The media type of the artifact (see [Model Artifact Media-Type](#model-artifact-media-type).     |
| roles             | \[string]                                 | **REQUIRED** Specify `mlm:model`. Can include `["mlm:weights", "mlm:checkpoint"]` as applicable. |
| mlm:artifact_type | [Artifact Type Enum](#artifact-type-enum) | Specifies the kind of model artifact. Typically related to a particular ML framework.            |

Recommended Asset `roles` include `mlm:weights` or `mlm:checkpoint` for model weights that need to be loaded by a
model definition and `mlm:compiled` for models that can be loaded directly without an intermediate model definition.
In each case, the `mlm:model` should be applied as well to indicate that this asset represents the model.

It is also recommended to make use of the
[file](https://github.com/stac-extensions/file?tab=readme-ov-file#asset--link-object-fields)
extension for this Asset, as it can provide useful information to validate the contents of the model definition,
by comparison with fields `file:checksum` and `file:size` for example.

#### Model Artifact Media-Type

> Not all ML framework, libraries or model artifacts provide explicit media-type. When those are not provided, custom

Reviewer: ```suggestion
Not all ML framework, libraries or model artifacts provide an explicit media-type. When those are not provided, custom
```

media-types can be considered. For example `application/x-pytorch` or `application/octet-stream; application=pytorch`
could be appropriate to represent a PyTorch `.pt` file, since the underlying format is a serialized pickle structure.

#### Artifact Type Enum

This value can be used to provide additional details about the specific model artifact being described.
For example, PyTorch offers various strategies for providing model definitions, such as Pickle (`.pt`), TorchScript,

> or the compiled approach. Since they all refer to the same ML framework,

Reviewer: ```suggestion
or the upcoming [Ahead-Of-Time Compiled .pt2 format](https://pytorch.org/docs/main/torch.compiler_aot_inductor.html). Since they all refer to the same ML framework,
```

the [Model Artifact Media-Type](#model-artifact-media-type) would be insufficient in this case to detect with strategy
should be used. 

Following are some proposed *Artifact Type* values for corresponding approaches, but other names are
permitted as well. Note that the names are selected using the framework-specific definitions to help
the users understand the source explicitly, although this is not strictly required either.

| Artifact Type      | Description                                                                                                              |
|--------------------|--------------------------------------------------------------------------------------------------------------------------|

> | `torch.compile`    | A model artifact obtained by [`torch.compile`](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html).  |

Reviewer: ```suggestion
| `.pt2`    | A model artifact obtained using [Pytorch's AOTInductor](https://pytorch.org/docs/main/torch.compiler_aot_inductor.html).  |
```

I think this is a necessary edit since torch.compile is an API for compiling pytorch nn.Modules. it's used to speed up Pytorch code in general, for training or inference. Many of the backend internals that make torch.compile work are used in AOTInductor (the tool that creates compiled model artifacts) but they aren't the same thing and I think here we want to refer to artifacts produced by AOTInductor

> | `torch.jit.script` | A model artifact obtained by [`TorchScript`](https://pytorch.org/docs/stable/jit.html).                                  |

Reviewer: ```suggestion
| `torchscript` | A model artifact obtained by [`TorchScript Scripting`](https://pytorch.org/docs/stable/jit.html) and/or [`TorchScript Tracing`](https://pytorch.org/docs/stable/generated/torch.jit.trace.html).                                  |
```

There are two types of graph capture in torchscript, trace and script. I think we can either enumerate both or leave only one option. Either traced or scripted models can be loaded the same way so I favor only one field for both of them.

Somewhat confusingly, both can also be used together though this is not common. https://ppwwyyxx.com/blog/2022/TorchScript-Tracing-vs-Scripting/

Since Torchscript is bring phased out in favor of AOTInductor, I think we shouldn't make this too complex and only provide them as one field.
